3rd-party media hosting category updates (#12227)
withastro/docs: src/content/docs/en/guides/media/mux.mdx

---
title: Mux & Astro
description: Add high performance video to your Astro project using Mux
sidebar:
  label: Mux
type: media
service: Mux
stub: true
i18nReady: true
---

import PackageManagerTabs from '~/components/tabs/PackageManagerTabs.astro';
import ReadMore from '~/components/ReadMore.astro';

[Mux](https://mux.com) is a hosted media service that provides video streaming infrastructure and performance analytics for businesses of all scale.

When you use Mux to store and host your video content, you'll have access to Astro-native video components for [Mux Player](#mux-player), a drop in component for adding Mux videos in your Astro project, and [Mux Uploader](#mux-uploader) for uploading videos to Mux from your website. These components integrate seamlessly with [Mux Data](https://www.mux.com/docs/guides/data) to track your video engagement and performance.

You can also interact with your content through the [Mux Node SDK](#mux-node-sdk).

:::tip

> Learn more about features such as embedding, storing, streaming, and customizing video at [Mux's dedicated page for video in Astro](https://www.mux.com/video-for/astro)!

Reviewer: Add offer code here?

Author: The only other place I considered it was in the "Prerequisites" section where we already direct people to sign up for a Mux account:

> - A Mux account. If you don’t have an account, you can [sign up with Mux](https://dashboard.mux.com/login) for a free account.

Strong preference for one vs. the other? It maybe feels more natural where we send people to log in or sign up than the tip, but I can understand why Mux might like it in a tip. Thoughts?

Author: Will put it in the prerequisites section!

:::

## Using Mux in Astro

Mux's APIs and web components work in Astro to compress and optimize your videos and streams for the web, adapt the quality of your video to network conditions, and integrate additional features like captions, thumbnails, and analytics. The [Mux Node SDK](https://github.com/muxinc/mux-node-sdk) supports both Mux Data and the Mux Video API.

## Prerequisites

- An existing Astro project. Some features may additionally require an adapter installed for [on-demand server rendering](/en/guides/on-demand-rendering/).
- A Mux account. If you don’t have an account, you can [sign up with Mux](https://dashboard.mux.com/login) using the code ASTRO to receive a $50 credit.

## Mux Player

In Astro, you can use the full-featured [Mux Player](https://www.mux.com/docs/guides/mux-player-web) as a native Astro component for optimized, responsive video playback and live streams. 

Mux Player provides a responsive UI based on video player dimensions and stream type, automatic thumbnail previews and poster images, and modern video player capabilities (e.g. fullscreen, picture-in-picture, Chromecast, AirPlay).

```astro title="src/components/MyMuxVideoPlayer.astro"
---

> import MuxUploader from "@mux/mux-player-astro";

Reviewer: ```suggestion
import MuxPlayer from "@mux/mux-player-astro";
```

---
<MuxPlayer
  playbackId="DS00Spx1CV902MCtPj5WknGlR102V5HFkDe"

>   metadata={{video_title: 'My Astro Video'}}

Reviewer: ```suggestion
  metadata={{ video_title: 'My Astro Video' }}
```

This will allow you to create an instance of the Mux Node SDK for retrieving information about your videos, creating new assets, accessing metrics and real-time performance, and more:

```astro title="src/components/StarlightVideo.astro"
---
import Mux from "@mux/mux-node";

const mux = new Mux ({
  tokenId: import.meta.env.MUX_TOKEN_ID,
  tokenSecret: import.meta.env.MUX_TOKEN_SECRET,
})
---
```

<ReadMore>Read more about using [environment variables](/en/guides/environment-variables/) in your Astro project, including creating a [type-safe schema](/en/guides/environment-variables/#type-safe-environment-variables) for your Mux credentials.</ReadMore>

### Retrieve asset data

To fetch information about your video to use in your Astro project, provide the video's `ASSET_ID` (available in the Mux dashboard) to the `retrieve()` helper function. This will allow you to pass values to both your Mux components and your HTML template, such as the video's title or duration:

```astro
---
import Mux from "@mux/mux-node";

> import MuxPlayer from "@mux/mux-player-astro";

Reviewer: ```suggestion
import { MuxPlayer } from "@mux/mux-player-astro";
```

const mux = new Mux ({
  tokenId: import.meta.env.MUX_TOKEN_ID,
  tokenSecret: import.meta.env.MUX_TOKEN_SECRET,
})

const ASSET_ID = "E01irAaN8c6dk1010153uC2mzst7RVbAdJJWtHECAHFvDo";
const asset = await mux.video.assets.retrieve(ASSET_ID);

const playbackId = asset.playback_ids?.find((id)=> id.policy=== "public")?.id;
const videoTitle = asset?.meta?.title;
const createdAt = Number(asset?.created_at);
const duration = Number(asset?.duration)

const date = new Date(createdAt * 1000).toDateString()
const time = new Date(Math.round(duration) * 1000).toISOString().substring(14, 19)
---
<h1>My Video Page</h1>
<p>Title: {videoTitle}</p>
<p>Upload Date: {date}</p>
<p>Length: {time}</p>

<MuxPlayer 
  playbackId={playbackId}
  videoTitle={videoTitle}
/>
```

See all asset properties in the [Mux Asset API documentation](https://www.mux.com/docs/api-reference/video/assets).

## Mux Uploader

>  [Mux Uploader](https://www.mux.com/docs/guides/mux-uploader) is a a fully-functional, customizable video upload UI for your Astro website. The native Astro `<MuxUpload />` component allows you to build video upload functionality in to your web app.

Reviewer 2: ```suggestion
[Mux Uploader](https://www.mux.com/docs/guides/mux-uploader) is a a fully-functional, customizable video upload UI for your Astro website. The native Astro `<MuxUpload />` component allows you to build video upload functionality in to your web app.
```

Mux Uploader supports both manual file selection and drag and drop for file uploads, optional pausing and resuming of uploads, and more.

### Installation

Install the Astro version of Mux Uploader using your preferred package manager:

<PackageManagerTabs>
  <Fragment slot="npm">
  ```shell
  npm install @mux/mux-uploader-astro
  ```
  </Fragment>
  <Fragment slot="pnpm">
  ```shell
  pnpm add @mux/mux-uploader-astro
  ```
  </Fragment>
  <Fragment slot="yarn">
  ```shell
  yarn add @mux/mux-uploader-astro
  ```
  </Fragment>
</PackageManagerTabs>

### Upload a video to Mux

> Before uploading a video, make sure you have your [Mux API access tokens](#mux-environment-api-access) configured. With those in place, you can use the create() function from the Mux Node SDK to start a new video upload:

Reviewer: ```suggestion
Before uploading a video, make sure you have your [Mux API access tokens](#mux-environment-api-access) configured. With those in place, you can use the `create()` function from the Mux Node SDK to start a new video upload:
```

```astro
---
import Layout from '../../layouts/Layout.astro';
import Mux from "@mux/mux-node";

> import MuxUploader from "@mux/mux-uploader-astro";

Reviewer: ```suggestion
import { MuxUploader } from "@mux/mux-uploader-astro";
```

service: Mux
stub: true
i18nReady: true
---

import PackageManagerTabs from '~/components/tabs/PackageManagerTabs.astro';
import ReadMore from '~/components/ReadMore.astro';

[Mux](https://www.mux.com?utm_campaign=21819274-Astro&utm_source=astro-docs) is a hosted media service that provides video streaming infrastructure and performance analytics for businesses of all scales.

When you use Mux to store and host your video content, you'll have access to Astro-native video components for [Mux Player](#mux-player), a drop-in component for adding Mux videos in your Astro project, and [Mux Uploader](#mux-uploader) for uploading videos to Mux from your website. These components integrate seamlessly with [Mux Data](https://www.mux.com/docs/guides/data?utm_campaign=21819274-Astro&utm_source=astro-docs) to track your video engagement and performance.

You can also interact with your content through the [Mux Node SDK](#mux-node-sdk).

:::tip
Learn more about features such as embedding, storing, streaming, and customizing video at [Mux's dedicated page for video in Astro](https://www.mux.com/video-for/astro?utm_campaign=21819274-Astro&utm_source=astro-docs)!
:::

## Using Mux in Astro

> Mux's APIs and web components work in Astro to compress and optimize your videos and streams for the web, adapt the quality of your video to network conditions, and integrate additional features like captions, thumbnails, and analytics. The [Mux Node SDK](https://github.com/muxinc/mux-node-sdk?utm_campaign=21819274-Astro&utm_source=astro-docs) supports both Mux Data and the Mux Video API.

Author: ```suggestion
Mux's APIs and web components work in Astro to compress and optimize your videos and streams for the web, adapt the quality of your video to network conditions, and integrate additional features like captions, thumbnails, and analytics. The [Mux Node SDK](https://www.mux.com/docs/integrations/mux-node-sdk?utm_campaign=21819274-Astro&utm_source=astro-docs) supports both Mux Data and the Mux Video API.
```

This will allow you to create an instance of the Mux Node SDK for retrieving information about your videos, creating new assets, accessing metrics and real-time performance, and more:

```astro title="src/components/StarlightVideo.astro"
---
import Mux from "@mux/mux-node";

const mux = new Mux ({
  tokenId: import.meta.env.MUX_TOKEN_ID,
  tokenSecret: import.meta.env.MUX_TOKEN_SECRET,
})
---
```

<ReadMore>Read more about using [environment variables](/en/guides/environment-variables/) in your Astro project, including creating a [type-safe schema](/en/guides/environment-variables/#type-safe-environment-variables) for your Mux credentials.</ReadMore>

### Retrieve asset data

To fetch information about your video to use in your Astro project, provide the video's `ASSET_ID` (available in the Mux dashboard) to the `retrieve()` helper function. This will allow you to pass values to both your Mux components and your HTML template, such as the video's title or duration:

```astro
---
import Mux from "@mux/mux-node";
import { MuxPlayer } from "@mux/mux-player-astro";

> const mux = new Mux ({

Reviewer 3: ```suggestion
const mux = new Mux({
```

```astro
---
import Layout from '../../layouts/Layout.astro';
import Mux from "@mux/mux-node";
import { MuxUploader } from "@mux/mux-uploader-astro";

const mux = new Mux({
  tokenId: import.meta.env.MUX_TOKEN_ID,
  tokenSecret: import.meta.env.MUX_TOKEN_SECRET
});

const upload = await mux.video.uploads.create({
  new_asset_settings: {
    playback_policy: ['public'],
    video_quality: 'basic'
  },
  cors_origin: '*',
});
---
<Layout title="Upload a video to Mux">
  <MuxUploader endpoint={upload.url} />
</Layout>
```

### Customize the uploader

You can customize the functionality and appearance of the `<MuxUploader />` with additional component attributes. In addition to styling your element, this allows you to control options such as the ability to pause a download or set a maximum file size.

```astro
---
import { MuxUploader } from '@mux/mux-uploader-astro';
---

<MuxUploader
  endpoint="https://my-authenticated-url/storage?your-url-params"
  pausable
  maxFileSize={1000000000}
  chunkSize={8192}
  style={{
    '--progress-bar-fill-color': '#7e22ce',
    '--button-background-color': '#f0f0f0',
  }}
/>
```

> See the [Mux uploader customization guide](https://www.mux.com/docs/guides/uploader-web-customize-look-and-feel?utm_campaign=21819274-Astro&utm_source=astro-docs) for more options.

Reviewer 3: ```suggestion
See the [Mux Uploader customization guide](https://www.mux.com/docs/guides/uploader-web-customize-look-and-feel?utm_campaign=21819274-Astro&utm_source=astro-docs) for more options.
```

<PackageManagerTabs>
  <Fragment slot="npm">
  ```shell
  npm install @mux/mux-node
  ```
  </Fragment>
  <Fragment slot="pnpm">
  ```shell
  pnpm add @mux/mux-node
  ```
  </Fragment>
  <Fragment slot="yarn">
  ```shell
  yarn add @mux/mux-node
  ```
  </Fragment>
</PackageManagerTabs>

### Mux Environment API access

API tokens are created for a specific Mux Environment: a group of videos for shared use and data analysis. You will need to create an Environment in your Mux dashboard, and [then get ID and SECRET tokens](https://www.mux.com/docs/core/stream-video-files#1-get-an-api-access-token) to provide credentials to the Node SDK. These can be provided to your Astro components as environment variables stored in a `.env` file.

This will allow you to create an instance of the Mux Node SDK for retrieving information about your videos, creating new assets, accessing metrics and real-time performance, and more:

```astro title="src/components/StarlightVideo.astro"
---
import Mux from "@mux/mux-node";

const mux = new Mux ({
  tokenId: import.meta.env.MUX_TOKEN_ID,
  tokenSecret: import.meta.env.MUX_TOKEN_SECRET,
})
---
```

> <ReadMore>Read more about using [environment variables](https://docs.astro.build/en/guides/environment-variables/) in your Astro project, including creating a [type-safe schema](/en/guides/environment-variables/#type-safe-environment-variables) for your Mux credentials.</ReadMore>

Author: ```suggestion
<ReadMore>Read more about using [environment variables](/en/guides/environment-variables/) in your Astro project, including creating a [type-safe schema](/en/guides/environment-variables/#type-safe-environment-variables) for your Mux credentials.</ReadMore>
```

const ASSET_ID = "E01irAaN8c6dk1010153uC2mzst7RVbAdJJWtHECAHFvDo";
const asset = await mux.video.assets.retrieve(ASSET_ID);

const playbackId = asset.playback_ids?.find((id)=> id.policy=== "public")?.id;
const videoTitle = asset?.meta?.title;
const createdAt = Number(asset?.created_at);
const duration = Number(asset?.duration)

const date = new Date(createdAt * 1000).toDateString()
const time = new Date(Math.round(duration) * 1000).toISOString().substring(14, 19)
---
<h1>My Video Page</h1>
<p>Title: {videoTitle}</p>
<p>Upload Date: {date}</p>
<p>Length: {time}</p>

<MuxPlayer 
  playbackId={playbackId}
  videoTitle={videoTitle}
/>
```

See all asset properties in the [Mux Asset API documentation](https://www.mux.com/docs/api-reference/video/assets).

## Mux Uploader

 [Mux Uploader](https://www.mux.com/docs/guides/mux-uploader) is a a fully-functional, customizable video upload UI for your Astro website. The native Astro `<MuxUpload />` component allows you to build video upload functionality in to your web app.

```astro title="src/components/MyMuxVideoUploader.astro"
---
import MuxUploader from "@mux/mux-uploader-astro";
---
<MuxUploader endpoint="https://httpbin.org/put" />

> ```

Author: removed!
